### ISSUE TITLE
Custom Catalog Product Image Role does not support Alt-Text (i.e. Label)

### ISSUE THREAD

**Issue by wilfriedwolf** [magento/magento2#34753]
### Preconditions (*)

1. Version 2.4.3
2. Define custom image role (e.g. "custom_image")
3. Define an Alt Text e.g. "Custom Alt Text" at some image with this role for some Product (e.g. with name "Product")
4. Initalize `\Magento\Catalog\Helper\Image` with store ID and entity ID.
5. Fetch Label with `$imageHelper->getLabel()`  

### Steps to reproduce (*)

Reproducible

### Expected result (*)

1. The result should be the defined Alt Text, e.g. "Custom Alt Text" 

### Actual result (*)

1. The result is the product name, e.g. "Product"

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by wilfriedwolf** [comment]
The Problem is that the list of media attributes with labels is Hardcoded here: `\Magento\Catalog\Model\Product\Gallery\CreateHandler::$mediaAttributesWithLabels`

@magento I am working on this

I'll submit a PR allowing DI to extend the list.

**Comment by wilfriedwolf** [comment]
@magento add to contributors team

**Comment by wilfriedwolf** [comment]
@magento add to contributors team

**Comment by wilfriedwolf** [comment]
The Problem is that the list of media attributes with labels is Hardcoded here: \Magento\Catalog\Model\Product\Gallery\CreateHandler::$mediaAttributesWithLabels

@magento I am working on this

I'll submit a PR allowing DI to extend the list.

**Comment by engcom-Hotel** [comment]
Hi @wilfriedwolf,

Thank you for reporting the issue:

I checked the issue description but the steps are not clear. I may have some idea what you are referring to but in order to confirm if it's an issue or Magento's expected behavior, I’ll have to reproduce this issue. The steps provided are not enough for me to reproduce this. In order to try and reproduce it, please provide more detailed steps to reproduce or perhaps add some images of the issue to the description. That would be helpful.

Also please explain, exactly where steps 3 and 4 are to be added or used?

Thanks

**Comment by wilfriedwolf** [comment]
Hi @engcom-Hotel ,

I'll provode you with a simple example in some, maybe tomorrow..


Best regards,

Willi

**Comment by wilfriedwolf** [comment]
Hi @engcom-Hotel 

I thougth providing you with a module, but it is simpler.

In vanilla Magento you have 3 Media Image Attributes (`image`, `small_image`, `thumbnail`) (Well actually 3, but `swatch_image` is a little different)

If you create a custom Media Image Attribute (`custom_image`) 

![image](https://user-images.githubusercontent.com/23700116/145218207-b5c76812-6655-476e-bc59-35e3ad79250c.png)

which defines a new Image Role which can be added to an attribute set:

![image](https://user-images.githubusercontent.com/23700116/145219890-3d57c856-c032-44bf-a235-6818f5455321.png)

and then attached to a product image:

![image](https://user-images.githubusercontent.com/23700116/145220629-b44eebe7-c054-40e4-8aa6-6372d5ba8bae.png)

Anyhow, if for some reason you need the Alt Text with this image role, you would never get it and that is for two reasons:

First there is no (hidden) attribute `custom_image_label` as there are `image_label`,`small_image_label`and `thumbnail_image`.

This can be worked arround by just creating one:

```
    /**
     * {@inheritdoc}
     */
    public function apply()
    {
        $this->moduleDataSetup->getConnection()->startSetup();
        /** @var EavSetup $eavSetup */
        $eavSetup = $this->eavSetupFactory->create(['setup' => $this->moduleDataSetup]);
        $eavSetup->addAttribute(
            Product::ENTITY,
            custom_image_label,
            [
                'type' => 'varchar',
                'label' => 'Custom Image Label',
                'input' => 'text',
                'required' => false,
                'sort_order' => 18,
                'global' => ScopedAttributeInterface::SCOPE_STORE,
                'visible' => false,
                'used_in_product_listing' => true,
            ]
        );

        $this->moduleDataSetup->getConnection()->endSetup();
    }
```

But the second problem is bigger: These "label attributes" get filled upon saving the Product in the `\Magento\Catalog\Model\Product\Gallery\CreateHandler` and there the `\Magento\Catalog\Model\Product\Gallery\CreateHandler::$mediaAttributesWithLabels` are hardcoded and not extensible by DI.

And yes you can work around this too, but it requires extending the `CreateHandler` and the `UpdateHandler`, copying lots of private stuff and using them by preference.

So my PR just proposes to allow passing an array of additional Media Image attributes to these two handlers so to simplyfy everything.

So hope that explains sufficiently.

Best regards,

Willi

**Comment by engcom-Echo** [comment]
Hi @wilfriedwolf,

Thank you for providing the details.

However I am not able to reproduce the ‘Alt Text’ issue. I tried with below steps:
1. Created 1 test product. 
2. Created Media Image Attribute with name ‘Custom Image’.
3. Added that attribute in Image set as you suggested.
4. Now added ‘Alt Text’ in Admin product image detail shown as below:

<img width="1268" alt="Screenshot 2021-12-27 at 1 37 59 PM" src="https://user-images.githubusercontent.com/51681435/147460432-0357c537-b367-4aca-ac7b-d8f69718dca3.png">

5. Refreshed cache, did reindexing and tried to find Alt text on frontend and I was able to locate it in Elements alongside image:

<img width="1074" alt="Screenshot 2021-12-27 at 1 38 46 PM" src="https://user-images.githubusercontent.com/51681435/147460454-3a1a5765-a0cc-45cc-b296-b5bfbd8c2942.png">

I am getting Alt Text with newly created Image Role on frontend so no issue there. 

Now, I think you are trying to get the same through code with Image Helper for which you are facing issue. So can you please provide below details for further analysis:
1. Sample code for understanding the issue better? I’ll be able to figure out if the issue is related to your custom code or related to Magento core functionality.
2. If you can provide 1 use case of the issue that you are facing like where exactly you require Alt text with custom built image role?

Thanks

**Comment by wilfriedwolf** [comment]
Hi @engcom-Echo 

I'll try to explain better:

Saving the image in the backend applying an Alt Text fills **all** the label attributes defined. See `\Magento\Catalog\Model\Product\Gallery\CreateHandler::processMediaAttributes`:

```
    /**
     * Update media attributes
     *
     * @param Product $product
     * @param array $existImages
     * @param array $newImages
     * @param array $clearImages
     */
    private function processMediaAttributes(
        Product $product,
        array $existImages,
        array $newImages,
        array $clearImages
    ): void {
        foreach ($this->getMediaAttributeCodes() as $mediaAttrCode) {
            $this->processMediaAttribute(
                $product,
                $mediaAttrCode,
                $clearImages,
                $newImages
            );
            if (in_array($mediaAttrCode, $this->mediaAttributesWithLabels)) {
                $this->processMediaAttributeLabel(
                    $product,
                    $mediaAttrCode,
                    $clearImages,
                    $newImages,
                    $existImages
                );
            }
        }
    }
```

So here for all media attributes (`image`,`small_image` ... `custom_image`) everything is saved.
But only for `$this->mediaAttributesWithLabels` the Label is saved.

The reason you can find the label is, that it is attached to all roles and on  the product detail page you display the `image` role, which has the Label Attached.

What you need to do, is to  create a block or template requesting the label for the `custom_image` role.
I actually do not have such code, because I use it in a custom REST API,
Anyhow you were right, it is because of 

`\Magento\Catalog\Helper\Image::init` 

This method allows you to fetch an Image of a certain role, but since fetching the label of this image in

`\Magento\Catalog\Helper\Image::getLabel` 

uses the beforeset `type` the fetching fails.

Anyhow, since in my understanding the label is attached to the image rather than to the role, I could use something like:

```
            if (!in_array($type, [
                'image',
                'small_image',
                'thumbnail'
            ])) {
                $product->setData($type.'_label', $product->getData('image_label'));
            }
```

in order to get the label for a specific image.

I'll try that out and close the issue in case it works.

Thank you for helping me better understandig the implementation.

Best regards,

Willi

**Comment by wilfriedwolf** [comment]
So finally I got to it.

Basically the problem was, that you can only get the Labels for (media gallery) images using

```
/** @var \Magento\Catalog\Helper\Image $imageHelper */
$imageHelper->init($product, $imageId);
$label = $imageHelper->getLabel();
```
if the image Id is in the 3 predifinend values (image, small_image or thumbnail), see above.

I now could solve it with a plugin on the init function above:

```
    /**
     * @param Image $subject
     * @param $result
     * @param Product $product
     * @param string $imageId
     * @param array $attributes
     */
    public function afterInit(Image $subject, $result, $product, $imageId, $attributes = [])
    {
        $type = $subject->getType();
        if (empty($product->getData($type.'_label'))) {
            try {
                $product->setData(
                    $type.'_label',
                    $product->getMediaGalleryImages()->getItemByColumnValue(
                        'file',
                        $product->getData($type)
                    ));
            } catch (\Throwable $exception) {
                ;
            }
        }
        return $result;
    }
```
which fills the data to be fetched by the `getLabel()`-Method.

Closing the ticket, thanks...
